posthog/management/commands/start_temporal_worker.py

```diff
@@ -149,27 +153,62 @@ def handle(self, *args, **options):
-            start_worker(
-                temporal_host,
-                temporal_port,
-                metrics_port=metrics_port,
-                namespace=namespace,
-                task_queue=task_queue,
-                server_root_ca_cert=server_root_ca_cert,
-                client_cert=client_cert,
-                client_key=client_key,
-                workflows=workflows,  # type: ignore
-                activities=activities,
-                graceful_shutdown_timeout=dt.timedelta(seconds=graceful_shutdown_timeout_seconds)
-                if graceful_shutdown_timeout_seconds is not None
-                else None,
-                max_concurrent_workflow_tasks=max_concurrent_workflow_tasks,
-                max_concurrent_activities=max_concurrent_activities,
+        shutdown_task = None
+
+        def shutdown_worker_on_signal(worker: Worker, sig: signal.Signals, loop: asyncio.events.AbstractEventLoop):
+            """Shutdown Temporal worker on receiving signal."""
+            nonlocal shutdown_task
+
+            logger.info("Signal %s received", sig)
+
+            if worker.is_shutdown:
+                logger.info("Temporal worker already shut down")
+                return
+
+            logger.info("Initiating Temporal worker shutdown")
+            shutdown_task = loop.create_task(worker.shutdown())
+            logger.info("Finished Temporal worker shutdown")
+
+        with asyncio.Runner() as runner:
+            worker = runner.run(
+                create_worker(
+                    temporal_host,
+                    temporal_port,
+                    metrics_port=metrics_port,
+                    namespace=namespace,
+                    task_queue=task_queue,
+                    server_root_ca_cert=server_root_ca_cert,
+                    client_cert=client_cert,
+                    client_key=client_key,
+                    workflows=workflows,  # type: ignore
+                    activities=activities,
+                    graceful_shutdown_timeout=dt.timedelta(seconds=graceful_shutdown_timeout_seconds)
+                    if graceful_shutdown_timeout_seconds is not None
+                    else None,
+                    max_concurrent_workflow_tasks=max_concurrent_workflow_tasks,
+                    max_concurrent_activities=max_concurrent_activities,
+                )
             )
-        )
+
+            loop = runner.get_loop()
+            for sig in (signal.SIGTERM, signal.SIGINT):
+                loop.add_signal_handler(
+                    sig,
+                    functools.partial(shutdown_worker_on_signal, worker=worker, sig=sig, loop=loop),
+                )
+                loop.add_signal_handler(
+                    sig,
+                    functools.partial(shutdown_worker_on_signal, worker=worker, sig=sig, loop=loop),
+                )
+
+            runner.run(worker.run())
+
+            if shutdown_task:
+                _ = runner.run(asyncio.wait([shutdown_task]))
```

Comment:
I'm not sure whether `run` is supposed to finish before or after `shutdown`, so I'm kind of checking both are done like this.